docs/source/clickhouse/schema_design.rst

```diff
@@ -11,36 +21,54 @@ and not load them for every row for every query is part of the performance advan
 ClickHouse provides over a traditional RDBMS (like PostgreSQL).
 
 Commonly, a data schema contains a flexible key:value pair mapping
-(canonically: ``tags`` or ``contexts``) and stores that
-data in a column that contains two ``Nested`` arrays where the first array contains the keys
-of the dictionary and the second array contains the values. This works well when
-your dataset design gives you the ability to filter based on other attributes to a small
-number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
-Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
-specific ``key=value`` pair. In this case one is often best-served by creating a top-level
-column and "promoting" the data to it [#dupe]_.
-
-This promotion is preferable on a couple dimensions:
-
-1. Even though it looks like a dictionary, our tags and contexts "maps" are actually arrays of keys and values.
-   With array accesses Clickhouse needs to locate the index of the key whose value needs to be looked up.
-   That is an O(N) lookup. The same key may be located at different indexes for each row.
-   Once the index is known, looking up its value is O(1) since you perform direct index access.
-   If the query has to lookup M rows, this becomes an O(M * N) in terms of time complexity.
-
-   With direct column access, there is no index lookup for a key. The overall complexity to lookup M rows
-   for "promoted columns" is O(M).
-2. Queries that filter based on an array column may have to load the entire column into memory
-   in order to do filtering. Adding a new index on top of a top-level column is a more
-   straightforward operation that guarantees more consistent performance outcomes vs. iterating
-   over an array
+(canonically at Sentry: ``tags`` or ``contexts``) and stores that
+data in a ``Nested`` column that contains two arrays where the first array contains the keys
+of the dictionary and the second array contains the values. To make queries faster,
+a column like this can be indexed with bloom filters as described in :ref:`bloom`. In general
+we construct this index across datasets for ``tags`` but not for other columns.
+
+This works well when your dataset and query design gives you the ability to
+filter for exact matches and a large number of rows will NOT be an exact match.
+Often, however, a ClickHouse query filters for rows that contain a substring match or regular
+expression match for a tag value of a given key. This makes bloom filter indexes
+not usable for the query and requires moving (or promoting) those relevant values for a given tag
```

Comment:
It does not "require" moving to a promoted column. 
This makes it seems like if we always promote a column in this case, which is not true. Most of the times we do not. 
Promotion makes things better but it is reserved for special cases.

Reply:
Tell me if my read on this is correct:

If your query pattern is low-selectivity *before* the inexact compare, it's fine to leave the tag un-promoted, because that will just be a filter in memory on a relatively small number of rows.

If your query pattern requires the inexact compare in order to be low-selectivity, then the promoted column becomes more necessary because ClickHouse will never finish iterating over all the arrays of a quasi table scan.